bindings/mqtt/mqtt.go

```diff
@@ -6,114 +6,258 @@
+		if !isValidPEM(val) {
+			return &m, fmt.Errorf("%s invalid client certificate key", errorMsgPrefix)
+		}
+		m.tlsCfg.clientKey = val
+	}
+
+	if val, ok := md.Properties[mqttBackOffMaxRetries]; ok && val != "" {
+		backOffMaxRetriesInt, err := strconv.Atoi(val)
+		if err != nil {
+			return &m, fmt.Errorf("%s invalid backOffMaxRetries %s, %s", errorMsgPrefix, val, err)
+		}
+		m.backOffMaxRetries = backOffMaxRetriesInt
+	}
+
+	return &m, nil
 }
 
-func (m *MQTT) getMQTTMetadata(metadata bindings.Metadata) (*mqttMetadata, error) {
-	b, err := json.Marshal(metadata.Properties)
+// Init does MQTT connection parsing
+func (m *MQTT) Init(metadata bindings.Metadata) error {
+	mqttMeta, err := parseMQTTMetaData(metadata)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	m.metadata = mqttMeta
 
-	var mMetadata mqttMetadata
-	err = json.Unmarshal(b, &mMetadata)
+	// mqtt broker allows only one connection at a given time from a clientID.
+	producerClientID := fmt.Sprintf("%s-producer", m.metadata.clientID)
+	p, err := m.connect(producerClientID)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return &mMetadata, nil
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	// TODO: Make the backoff configurable for constant or exponential
+	b := backoff.NewConstantBackOff(5 * time.Second)
+	m.backOff = backoff.WithContext(b, m.ctx)
+
+	m.producer = p
+
+	m.logger.Debug("mqtt message bus initialization complete")
+
+	return nil
 }
 
 func (m *MQTT) Operations() []bindings.OperationKind {
 	return []bindings.OperationKind{bindings.CreateOperation}
 }
 
 func (m *MQTT) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
-	m.client.Publish(m.metadata.Topic, 0, false, string(req.Data))
-	m.client.Disconnect(0)
+	m.logger.Debugf("mqtt publishing topic %s with data: %v", m.metadata.topic, req.Data)
+
+	token := m.producer.Publish(m.metadata.topic, m.metadata.qos, m.metadata.retain, req.Data)
+	if !token.WaitTimeout(defaultWait) || token.Error() != nil {
+		return nil, fmt.Errorf("mqtt error from publish: %v", token.Error())
+	}
 
 	return nil, nil
 }
 
 func (m *MQTT) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
-	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	sigterm := make(chan os.Signal, 1)
+	signal.Notify(sigterm, os.Interrupt, syscall.SIGTERM)
+
+	// reset synchronization
+	if m.consumer != nil {
+		m.logger.Warnf("re-initializing the subscriber")
+		m.consumer.Disconnect(0)
+		m.consumer = nil
+	}
+
+	// mqtt broker allows only one connection at a given time from a clientID.
+	consumerClientID := fmt.Sprintf("%s-consumer", m.metadata.clientID)
+	c, err := m.connect(consumerClientID)
+	if err != nil {
+		return err
+	}
+	m.consumer = c
+
+	m.logger.Debugf("mqtt subscribing to topic %s", m.metadata.topic)
+	token := m.consumer.Subscribe(m.metadata.topic, m.metadata.qos, func(client mqtt.Client, mqttMsg mqtt.Message) {
```

Comment:
Is this coming from the pubsub implementation?

Reply:
Yes it's coming from the MQTT pubsub component